core/system/src/system.c

```diff
@@ -16,16 +16,46 @@
  */
 #include <uvisor.h>
 
-/* All system IRQs are by default weakly linked to the system default handler */
+/* All non-hookable system IRQs are by default weakly linked to the system
+ * default handler. */
 void UVISOR_ALIAS(isr_default_sys_handler) NonMaskableInt_IRQn_Handler(void);
 void UVISOR_ALIAS(isr_default_sys_handler) HardFault_IRQn_Handler(void);
 void UVISOR_ALIAS(isr_default_sys_handler) MemoryManagement_IRQn_Handler(void);
 void UVISOR_ALIAS(isr_default_sys_handler) BusFault_IRQn_Handler(void);
 void UVISOR_ALIAS(isr_default_sys_handler) UsageFault_IRQn_Handler(void);
 void UVISOR_ALIAS(isr_default_sys_handler) SVCall_IRQn_Handler(void);
 void UVISOR_ALIAS(isr_default_sys_handler) DebugMonitor_IRQn_Handler(void);
-void UVISOR_ALIAS(isr_default_sys_handler) PendSV_IRQn_Handler(void);
-void UVISOR_ALIAS(isr_default_sys_handler) SysTick_IRQn_Handler(void);
+
+/* Privileged PendSV and SysTick hooks assume that they are called as the by
+ * the hardware. So, these handlers need their stack frame and registers to
+ * look exactly like they were called directly by hardware. This means LR needs
+ * to be EXC_RETURN, not some value placed there by the C compiler when it does
+ * a branch with link.
+ *
+ * These handlers are UVISOR_NORETURN so that the C call of the hook happens
+ * with a branch that doesn't affect LR.
+ *
+ * These handlers are marked as naked because we don't need the C pre-amble
+ * (which saves the LR and R7 on the stack). */
+void UVISOR_NAKED PendSV_IRQn_Handler(void)
+{
+    asm volatile(
+        "ldr  r0, =g_priv_sys_irq_hooks\n" // Load the address of the hook table.
```

Comment:
Could you make these comments `/* ... */`-style? I've been trying to enforce this in my latest asm, and I'll fix old asm at some point. It's more robust (the asm strings are concatenated and some compilers might make weird things with the `// ...` comments) and it's also MISRA-compliant.